### PR TITLE
Components - Upgraded the TFX components to 0.21.4

### PR DESCRIPTION
**Pull request by Ark-kun**
Updated and synced the generated code

There is only 1 line of component specific code in each component function (apart from the function signature). `from tfx.components.evaluator.component import Evaluator as component_class`

* Updated some components that had older version of the generated code. The generated code is now the same everywhere.
* `input_channels_with_splits` is now generated based on the input artifact types
* TFX broke back compat: Removed `.split` from the artifacts. The components seem to now assume there is a single artifact in the channel.
* TFX broke back compat: changed the way artifact instances are created

Updated component signatures to v0.21.4. Many of the components have changed the input and output names or got extra inputs/outputs.

Updated the sample pipeline.

### CODE REVIEW

**Comment by Bobgy** [comment]
a rough idea, if we can set up periodic cron jobs that create PRs like this, it will be a lot less effort to keep them in sync

Does it make sense?

**Comment by Ark-kun** [comment]
> a rough idea, if we can set up periodic cron jobs that create PRs like this, it will be a lot less effort to keep them in sync
> 
> Does it make sense?

Good idea. I want to add a script to generate these components automatically.
Unfortunately, TFX often breaks backwards compatibility so the code will need to be updated when this happens.

File: components/tfx/Transform/component.py [hunk 1]
```diff
@@ -4,16 +4,15 @@
...
 def Transform(
-    input_data_path: InputPath('Examples'),
-    #examples: InputPath('Examples'),
+    examples_path: InputPath('Examples'),
     schema_path: InputPath('Schema'),
 
-    transform_output_path: OutputPath('TransformGraph'),
-    #transform_graph_path: OutputPath('TransformGraph'),
+    transform_graph_path: OutputPath('TransformGraph'),
     transformed_examples_path: OutputPath('Examples'),
```
Comment (rmgogogo) [inline components/tfx/Transform/component.py:11]: this one can be merged with examples_path

File: components/tfx/Trainer/component.py [hunk 2]
```diff
@@ -5,18 +5,19 @@
...
-    #transform_graph_path: InputPath('TransformGraph'),
     schema_path: InputPath('Schema'),
 
-    output_path: OutputPath('Model'),
+    model_path: OutputPath('Model'),
 
+    train_args: {'JsonObject': {'data_type': 'proto:tfx.components.trainer.TrainArgs'}},
+    eval_args: {'JsonObject': {'data_type': 'proto:tfx.components.trainer.EvalArgs'}},
     module_file: str = None,
     trainer_fn: str = None,
```
Comment (rmgogogo) [inline components/tfx/Trainer/component.py:15]: will you have another CL to address the changes for native Keras (which will replace Estimator?)

Or name it as "run_fn" in this PR? May prefer you split it to another PR.

File: components/tfx/Transform/component.py (hunk 1 above)
Comment (Ark-kun) [reply components/tfx/Transform/component.py:11]: All of these parameters correspond to the parameters of TFX components.
`examples_path` points to the source examples while the `transformed_examples_path` is for the transformed examples.

File: components/tfx/Trainer/component.py (hunk 2 above)
Comment (Ark-kun) [reply components/tfx/Trainer/component.py:15]: These components just call the existing TFX components. When TFX components switch to Keras, these components will also switch.

**Comment by Ark-kun** [comment]
Thanks for the review.

> I thought TFX user will just use TFX SDK with Kubeflow runner.

That's true. But we also have KFP users who build their pipelines using KFP SDK and who use KFP components (I see several hundred on GitHib). It's strange to have components for XGBoost, but not Tensorflow. Previously, when we deleted the old TFX components we started getting user requests for TFMA, TFDV etc.

>feel free to submit in your timezone. I leave few comments for Native Keras. If this is to be maintained in long run, better support Native Keras

Since these components just call the official TFX components, when TFX components switch to Keras, these components will switch too.
